installinator/src/reporter.rs

```diff
@@ -49,129 +78,418 @@ impl ProgressReporter {
+                    // as well.
+                    slog::info!(
                         log,
-                        "failed to discover peers: {}",
-                        DisplayErrorChain::new(error.as_error()),
+                        "cancellation received, cancelling discover task",
                     );
-                    return None;
+                    // It's fine to drop discover_task_loop at this point, since
+                    // the only writes it does are publications to the peers_tx
+                    // watch channel.
                 }
-                #[cfg(test)]
-                Err(DiscoverPeersError::Abort(_)) => {
-                    // This is not possible, since test implementations don't
-                    // currently generate Abort errors during reporter
-                    // discovery.
-                    unreachable!(
-                        "DiscoverPeersError::Abort is not generated for the \
-                         reporter by test implementations"
-                    )
+            }
+        })
+    }
+
+    /// Spawn a task to listen to lists of peers, and:
+    ///
+    /// * spawn new report tasks for peers that are new
+    /// * cancel report tasks for peers that are no longer present
+    ///
+    /// This task is persistent and activates on changes to `peers_rx`.
+    fn spawn_reconcile_peers_task(
+        &self,
+        mut peers_rx: watch::Receiver<Option<PeerAddresses>>,
+        report_rx: watch::Receiver<Option<ReportMessage>>,
+    ) -> JoinHandle<()> {
+        let log = self.log.new(slog::o!("task" => "reconcile_peers"));
+        let update_id = self.update_id;
+        let backend = self.report_backend.clone();
+        // A map of peers to their report tasks.
+        let mut peer_tasks: BTreeMap<PeerAddress, ReportTask> = BTreeMap::new();
+
+        tokio::spawn(async move {
+            loop {
+                // Listen to changes in the list of peers.
+                let peers = match peers_rx.changed().await {
+                    Ok(()) => peers_rx.borrow_and_update(),
+                    Err(error) => {
+                        slog::warn!(
+                            log,
+                            "failed to receive peers (sender dropped, \
+                             discovery task likely completed), exiting: {}",
+                            error
+                        );
+                        return;
+                    }
+                };
+
+                let Some(peers) = peers.as_ref() else {
+                    continue;
+                };
+
+                let before_peers =
+                    peer_tasks.keys().cloned().collect::<BTreeSet<_>>();
+                let after_peers = peers.peers().clone();
+                let diff = before_peers.diff(&after_peers);
+
+                // For each peer in diff.removed, cancel the report task.
+                for peer in diff.removed {
```

Comment:
This is a really nice use of daft!

Reply:
Yeah, I've become a big fan of daft just for doing map and set difference computations.